### ISSUE TITLE
MPI data preparation

### ISSUE THREAD

**Issue by pierrethibault** [ptycho/ptypy#51]
Data preparation sometime fails when running with MPI. We should either debug this or consider falling back to single-process data preparation.

**Comment by alexbjorling** [comment]
I'm having the issue that _mpi_save_chunk() fails to MPI-send data and instead raises the ValueError "ndarray is not contiguous". It works if you modify utils.parallel.send() such that it sends the contiguous copy np.ascontiguousarray(data) of the array, but I guess that this is not the proper solution.

Is this the failure you are referring to, @pierrethibault ?

**Comment by pierrethibault** [comment]
@stefanoschalkidis Can you confirm that this is also the error message you get? If so, I think that making a contiguous copy is a good fix for now.

**Comment by stefanoschalkidis** [comment]
I'm currently getting a different .h5 error, but this is probably due to the messy structure of the I13 subclass. I will try to confirm the above once I've reorganized I13.

**Comment by alexbjorling** [comment]
I'm getting this error both with some I13 far-field data that Aaron kindly sent me, and with data taken here at NanoMax and loaded with my own PtyScan subclass. It happens as soon as I run on more than one process.

**Comment by stefanoschalkidis** [comment]
Yes, I think I got a similar error with I13 near-field data, but then I introduced some switches to the subclass and now the code breaks earlier due to an .h5 error when run with more than one core. I intend to reorganize the script for near-field soon and will then try to recreate the original behaviour.

**Comment by stefanoschalkidis** [comment]
Hi all, I'm done with reorganising the I13 NFP subclass and now I don't get any error when loading with MPI. The code however breaks later in classes.py, but the prepared .ptyd file is perfectly fine, meaning it will work when the reconstruction is loaded directly from there (also with MPI). I will try to look into the new error, but for now I don't think I can comment on the _mpi_save_chunk() failure.

**Comment by stefanoschalkidis** [comment]
Hi, out of curiosity I also reorganised the I13 FFP subclass and now I get the same error. The contiguous array fixes this issue, but then the code stops at the hard coded megapixel limit. @alexbjorling do you observe the same behaviour?

**Comment by alexbjorling** [comment]
No I don't, but maybe you have a larger dataset than I do.

@pierrethibault, is this the same error that you refer to in the original issue?

**Comment by pierrethibault** [comment]
I filed the bug, but it was observed by Stefanos. @stefanoschalkidis I think the megapixel error is another one - it is almost always caused by wrong scaling of positions vs pixel size. Since this seems to happen often I suggest that we make the error message more informative (dump the geometry parameters associated with the storage would do the trick).

**Comment by bjoernenders** [comment]
@alexbjorling could you please share the subclass with us, that is having trouble with "_mpi_save_chunk". It could be an implementation error.

**Comment by alexbjorling** [comment]
Yep, here it is. Excuse the very many personal notes... Note that I'm making a contiguous copy of the hdf5 object slice, just in case. Maybe this loading is MPI-incompetent for some other reason?

```
"""  Implementation of a PtyScan subclass to hold nanomax scan data. The
format is very preliminary, and therefore the class NanomaxTmpScan  is subject
to frequent change.  """

import ptypy
from ptypy.core.data import PtyScan
from ptypy.utils import Param
import h5py
import numpy as np

# the PtyScan class takes a parameter tree scans.xxx.data as  its constructor
# argument. These are default values of that tree, which include the default
# recipe.

# All "additional" parameters must come from the recipe tree, and the recipe
# is filled in from the script where data preparation is initiated. Here's the
# default.
RECIPE = Param()
RECIPE.dataPath = None
RECIPE.datafile = None
RECIPE.maskfile = None
RECIPE.maxFrames = None     # read only the first N frames
RECIPE.stepsize = None      # step size in m

# These are the paths within hdf5 files
NEXUS_DATA_PATH = 'entry/detector/data'

class NanomaxTmpScan(PtyScan):

    def __init__(self, pars=None, **kwargs):

        # Get the default parameter tree and add a recipe branch. Here, the
        # parameter tree is that corresponding to a scan.data tree, I guess it
        # will eventually become a scans.xxx.data tree.
        p = PtyScan.DEFAULT.copy()
        p.recipe = RECIPE.copy()
        p.update(pars, in_place_depth=10)
        p.update(kwargs)

        # Call the base class constructor with the new updated parameters.
        # This constructor only reads params, it doesn't modify them. We've
        # already put the kwargs in p, so we don't need to pass them here.
        super(NanomaxTmpScan, self).__init__(p)

        # From now on, all the input parameters are available through self.info.
        # Some stuff is available also through self.meta.

        # should probably instantiate Geo here somewhere and get distance,
        # psize, shape from somewhere.

    def load_positions(self):  
        # This is based on a very early hdf5 format and assumes that the scan
        # is sqrt(Nframes)-by- sqrt(Nframes).

        # load_positions causes self.num_frames to get set, so here we can
        # limit how many files to read I think.

        # The docs say that positions become available as self.positions, but
        # instead it's in self.info.positions_scan.

        filename = self.info.recipe.dataPath + self.info.recipe.datafile
        with h5py.File(filename, 'r') as hf:
            data = hf.get(NEXUS_DATA_PATH)
            data = np.array(data)
            self.nPositions = data.shape[0]
        positions = []
        for motor_y in range(int(np.sqrt(self.nPositions))):
            # tried to adapt this to Bjoern's info here
            # https://github.com/ptycho/ptypy-dev/issues/39
            for motor_x in range(int(np.sqrt(self.nPositions))):
                positions.append(np.array([-motor_y, -motor_x]) * self.info.recipe.stepsize)
                if len(positions) == self.info.recipe.maxFrames:
                    return np.array(positions)

    def load(self, indices):
        # returns three dicts: raw, positions, weights, whose keys are the
        # scan pont indices. If one weight (mask) is to be used for the whole
        # scan, it should be loaded with load_weights(). The same goes for the
        # positions. We don't really need a mask here, but it must be
        # provided, otherwise it's given the shape of self.info.shape, and
        # then there's a shape mismatch in some multiplication.

        # Probably this should slice up the big array on reading, but won't bother now. 

        raw, weights, positions = {}, {}, {}

        filename = self.info.recipe.dataPath + self.info.recipe.datafile
        with h5py.File(filename) as hf:
            data = hf.get(NEXUS_DATA_PATH)
            for i in indices:
                raw[i] = np.ascontiguousarray(data[i])
                weights[i] = np.ones(data[i].shape)

        return raw, positions, weights
```

**Comment by stefanoschalkidis** [comment]
@alexbjorling could you maybe tell me where your code stops when you use MPI for loading? Many thanks

**Comment by alexbjorling** [comment]
```
ValueErrorTraceback (most recent call last)
/home/alex/ptycho/ptypyNanomax/zoneplatescan_reconstruct.py in <module>()
    154 p.engines.engine_00.numiter = 10
    155 
--> 156 P = Ptycho(p,level=5)
    157 
    158 import time

/home/alex/.local/lib/python2.7/site-packages/ptypy/core/ptycho.pyc in __init__(self, pars, level, **kwargs)
    177         if level >= 2:
    178             logger.info('\n' + headerline('Ptycho init level 2', 'l'))
--> 179             self.init_data()
    180         if level >= 3:
    181             logger.info('\n' + headerline('Ptycho init level 3', 'l'))

/home/alex/.local/lib/python2.7/site-packages/ptypy/core/ptycho.pyc in init_data(self, print_stats)
    322         # Load the data. This call creates automatically the scan managers,
    323         # which create the views and the PODs.
--> 324         self.modelm.new_data()
    325 
    326         # Print stats

/home/alex/.local/lib/python2.7/site-packages/ptypy/core/manager.pyc in new_data(self)
    305         # For some funny reason the Generator constuct used to fail.
    306         while True:
--> 307             dp =  self.ptycho.datasource.feed_data()
    308             if dp is None: break
    309             """

/home/alex/.local/lib/python2.7/site-packages/ptypy/core/data.pyc in feed_data(self)
   1456             cur_scan.initialize()
   1457 
-> 1458         msg = cur_scan.auto(self.frames_per_call, self.feed_format)
   1459 
   1460         # If we catch a scan that has ended, look for an unfinished scan

/home/alex/.local/lib/python2.7/site-packages/ptypy/core/data.pyc in auto(self, frames, chunk_form)
    797             # save chunk
    798             if self.info.save is not None:
--> 799                 self._mpi_save_chunk(self.info.save, msg)
    800             # delete chunk
    801             del self.chunk

/home/alex/.local/lib/python2.7/site-packages/ptypy/core/data.pyc in _mpi_save_chunk(self, kind, chunk)
   1081                 parallel.barrier()
   1082                 # Gather the content
-> 1083                 newv = parallel.gather_dict(v)
   1084                 todisk[k] = np.asarray([newv[j] for j in sorted(newv.keys())])
   1085 

/home/alex/.local/lib/python2.7/site-packages/ptypy/utils/parallel.pyc in gather_dict(dct, target)
    552                 #send(k, dest=target)
    553                 comm.send(k, dest=target,tag=9999)
--> 554                 send(v, dest=target)
    555         barrier()
    556 

/home/alex/.local/lib/python2.7/site-packages/ptypy/utils/parallel.pyc in send(data, dest, tag)
    311             comm.Send(data.astype('|u1'), dest=dest, tag=tag)
    312         else:
--> 313             comm.Send(data, dest=dest, tag=tag)
    314     else:
    315         # Send pickled whatever thing

MPI/Comm.pyx in mpi4py.MPI.Comm.Send (src/mpi4py.MPI.c:92870)()

MPI/msgbuffer.pxi in mpi4py.MPI.message_p2p_send (src/mpi4py.MPI.c:33688)()

MPI/msgbuffer.pxi in mpi4py.MPI._p_msg_p2p.for_send (src/mpi4py.MPI.c:33550)()

MPI/msgbuffer.pxi in mpi4py.MPI.message_simple (src/mpi4py.MPI.c:30349)()

MPI/msgbuffer.pxi in mpi4py.MPI.message_basic (src/mpi4py.MPI.c:29359)()

MPI/asbuffer.pxi in mpi4py.MPI.getbuffer (src/mpi4py.MPI.c:9310)()

MPI/asbuffer.pxi in mpi4py.MPI.PyObject_GetBufferEx (src/mpi4py.MPI.c:8070)()

ValueError: ndarray is not contiguous
```

**Comment by stefanoschalkidis** [comment]
Thanks! I assume you've tried this already:

`--> 313             comm.Send(data, dest=dest, tag=tag)`
to
`--> 313             comm.Send( np.ascontiguousarray(data), dest=dest, tag=tag)`

Does the code stop anywhere after applying this?

**Comment by alexbjorling** [comment]
Exactly, that fixes the problem as I mentioned in my first comment above. It's a workaround, but is it an acceptable longer-term fix, @bjoernenders? I mean it adds some memory usage and that might have consequences somewhere else where big arrays are sent.

Seems to me that if the current comm.Send() call works for some PtyScan subclasses, then we should find out how these have to be written to work.

**Comment by pierrethibault** [comment]
If I may pitch in: @alexbjorling could you run the code with a debugger and tell what is the `data` that is being sent in `comm.Send`? Maybe we will be surprised. Otherwise, `numpy.ascontiguous` does not copy if the data is already contiguous, so I would guess there is little overhead.

**Comment by alexbjorling** [comment]
Sure. The data being sent is a detector image, a numpy array of `shape` `(128, 128)` and of `dtype` `int32`. The flags are like this:

```
C_CONTIGUOUS : False
F_CONTIGUOUS : False
OWNDATA : False
WRITEABLE : True
ALIGNED : True
UPDATEIFCOPY : False
```

So that's obviously not good. But as you can see in my `PtyScan` subclass above, I do `np.ascontiguousarray` and I think that my `load()` function returns the data as specified in the `PtyScan` docs. Maybe I'm confused, but is seems like the data are being scrambled somewhere between my `PtyScan` subclass constructor and the `comm.Send()` call.

**Comment by bjoernenders** [comment]
I would suspect the .get() command of h5py is the culprit here. @alexbjorling have you tried replacing the data = hf. get() part with just a regular fresh cast of a numpy array like data = np.ones((number_of_frames,128,128))? If the error persists, there is something wrong internally. Otherwise your reimplementation of load looks ok.

However in load_position I wouldn't advise to hide the 'return' statement in and if-clause in a for loop. That can lead to unexpected behavior in future.

**Comment by bjoernenders** [comment]
@alexbjorling Did you try what I suggested. Does the bug still persist?

**Comment by alexbjorling** [comment]
Sorry, forgot to try. No, with a fresh numpy array the error persists.

```
    def load(self, indices):
        raw, weights, positions = {}, {}, {}

        filename = self.info.recipe.dataPath + self.info.recipe.datafile
        with h5py.File(filename) as hf:
            #data = hf.get(NEXUS_DATA_PATH)
            data = np.ones((2601, 195, 487), dtype=np.int)
            for i in indices:
                raw[i] = np.asarray(data[i])
                weights[i] = np.ones(data[i].shape)

        return raw, positions, weights
```

produces 

```
ValueError: ndarray is not contiguous
```

as above.

**Comment by bjoernenders** [comment]
Ok Alex this seems very strange to me now. Could you share the nexus file with me, then I try on my own. Maybe its mpi4py.

**Comment by alexbjorling** [comment]
Sure thing, I've sent you a Dropbox share with the hdf5 file and the scipt I'm using, and pushed my "nanomax" branch to this repo. 

This breaks:

```
mpiexec -n 2 python zoneplatescan_reconstruct.py
```

This works:

```
mpiexec -n 1 python zoneplatescan_reconstruct.py
```

**Comment by alexbjorling** [comment]
Ok so the problem is that MPI (at least in C but apparently also in python) just sends data chunks straight from memory. You can define derived MPI types to send for example a column of a C-array, but it's tricky.

In numpy, assigning an array by slicing another one usually produces a discontiguous array, as happens in `u.crop_pad_symmetric_2d()`. Transposing a C-contiguous (row-wise) array of course produces an F-contiguous (column-wise) one, etc.

The bottom line is that all the crop/pad/rebin/rotate/flip operations cause discontiguity. So added a call to `np.ascontiguousarray()` in data.py where this happens. Not sure if it's better to have this done in the `Send()` wrapper as above, but my fix is closer to the root of the problem.

EDIT: @bjoernenders, the master branch is protected, so can't push.

**Comment by bjoernenders** [comment]
Hi Alex, hmm does your push not automatically transform into a pull request? Otherwise just issue one. I'll try to stay more alert now. Alternatively (and that is more canonical, you should develop on a feature / bug branch) I can pull in these changes very easily. I'll stay more up-to-date this time.
I do no longer want people to be able to blindly push to master, so I keep it protected. Sorry for the inconvenience.

**Comment by bjoernenders** [comment]
Ok bug is merged. Does anyone else still have this issue? @pierrethibault 
Otherwise I close this issue.

**Comment by alexbjorling** [comment]
All my PtyScan subclasses work in MPI after the fix, I think the issue can go.

**Comment by simonesala** [comment]
Hi all,
This issue is still unsolved for I13 data and I can't seem to find what's causing the problem.
As already known, the code runs fine without mpi or with 'mpiexec -n 1' but as soon as another core is introduced while loading with I13's recipe (i.e. rather than 'file') it stops working.
Data loading is successful and the initialisation of the first process too (in POD creation): the second process spits out the following error and the code gets stuck, i.e. doesn't break which I'm guessing is due to the other process just hanging and waiting.

Traceback (most recent call last):
  File "ptypy_95103_14_prep.py", line 190, in <module>
    P = Ptycho(p, level=4)
  File "/home/sala/.local/lib/python2.7/site-packages/ptypy/core/ptycho.py", line 193, in __init__
    self.init_data()
  File "/home/sala/.local/lib/python2.7/site-packages/ptypy/core/ptycho.py", line 338, in init_data
    self.modelm.new_data()
  File "/home/sala/.local/lib/python2.7/site-packages/ptypy/core/manager.py", line 607, in new_data
    self.ptycho.obj.reformat(True)
  File "/home/sala/.local/lib/python2.7/site-packages/ptypy/core/classes.py", line 1662, in reformat
    s.reformat()
  File "/home/sala/.local/lib/python2.7/site-packages/ptypy/core/classes.py", line 656, in reformat
    'requested %.2fM pixels.' % megapixels)
RuntimeError: Arrays larger than 50M not supported. You requested 1273.74M pixels.

This is run from the latest master branch, so the 'ascontiguous' patch's been long introduced. Also it's not a scaling error directly caused by input geometry (e.g. pixel size) as the first process is initialised smoothly and the whole thing runs without issues in all other conditions (i.e. loading from file and/or on one core only).

It occurs for both 'i13dls_ffp' and 'i13dls_nfp'.
With 'id16a_nfp' everything works fine.

**Comment by aaron-parsons** [comment]
Which branch is it on? And specifically can you send the script you are running and the dataset you are running it on?

I’ll try to take a look this afternoon.

Usually when I see this error, it means the pixel size in realspace is being set to tiny which causes the arrays to be huge. Are you sure the motors are in the correct (metres) units, and that all the other distances, energies and wavelengths are in SI. Sorry if that sounds pedantic, it’s just happened to me now so many times…☺

**Comment by simonesala** [comment]
Hi,
As I said, this is run on the latest master branch and geometry is unlikely to be the main reason for failure as running the same script without mpi or with mpiexec -n 1 generates no error.
Anyway I uploaded data and script on a google drive (https://drive.google.com/drive/folders/0B234_PaRxpgsTG1IWGhwWTJGLTA?usp=sharing): you'll just have to modify your base_path at the very beginning of the script before running and this should reproduce the error.
Thanks for looking into this!
S

**Comment by aaron-parsons** [comment]
Hi,

Looking at this now.

Can you send me your probe file? That way I don’t have to change your script!

Aaron

**Comment by aaron-parsons** [comment]
OK. I can reproduce this on the dls branch now, so you are not alone!

Looking into it now.

**Comment by aaron-parsons** [comment]
I should also mention that I reproduce it in a different ptyscan subclass –dls. I’ve run out of time for today, but will revisit tomorrow afternoon.

From: Parsons, Aaron (DLSLtd,RAL,SCI)
Sent: 01 June 2017 16:49
To: 'ptycho/ptypy'; ptycho/ptypy
Cc: Comment
Subject: RE: [ptycho/ptypy] MPI data preparation (#51)

OK. I can reproduce this on the dls branch now, so you are not alone!

Looking into it now.

**Comment by aaron-parsons** [comment]
@darrenbatey will look into this further. We will also talk about it with Bjoern over the coming days.

**Comment by simonesala** [comment]
I do not consider the following a fix - so I will not close the issue.
It's been pointed out to me that using load_parallel = 'all' rather than load_parallel = 'data' in the parameter tree within the launching script makes the issue disappear, i.e. everything is loaded smoothly with several cores and reconstructions run as expected.
This could then be a quick fix/check to keep in mind for now.
Some more looking into this will be needed before deciding whether such behaviour is intended which I do not foresee having time for in the near future.

**Comment by alexbjorling** [comment]
Right, since this is showing up again in other contexts (like #153) maybe we'd better just do `Send(np.ascontiguousarray(data))` in the MPI code. My fix above fixed the discontiguities caused by the crop/pad code, but in principle other operations in data preparation (even slicing) could cause this too.
